04715 Implement the modular Pre-Handle Workflow

rbair23 opened the pull request:
Closes #4715.

This PR represents a complete implementation of the pre-handle workflow, as far as we understand it today. It includes tests and some primitive JMH benchmarks.

**JMH**
- Upgrade JMH to the latest release
- Setup `hedera-app` with JMH in the build
- Introduce the start of a JMH test for signature verification and signature expansion. Further refinements and new tests in the future.

**`hedera-app` build**
- Create test fixtures for `hedera-app`

**Signature Verification**
- Introduce a new and complete implementation of signature verification with no reliance on the mono-service.
- SignatureVerifier interface is the only interaction between the pre-handle workflow and the signature verification facility
- Supports hollow account based verification and key based verification
- Deprecates `SigExpansionResult`, `SignaturePreparer`, etc.

**Record Cache**
- Some basic plumbing for `RecordCache`. Not fully right. More to come in the future.
- Remove TODO from `SubmissionManager` by submitting the receipt into the `RecordCache`

**Handle Context**
- Add to handle context some helper methods for getting verification results by key and account ID (for hollow accounts). Probably not the right API, will be refined by the handle-workflow PRs.

**Pre-Handle workflow**
- Refactor heavily pre-handle workflow and `PreHandleResult`
- Support key and hollow account based payers and non-payers
- These classes are modified enough to consider it a rewrite -- the diff is probably not helpful, but it is worth reading straight.

**Scenario Builders**
For tests, I used a combination of mocking and convenience APIs for building real accounts and transactions. It kept the test logic closer to the essence.

rbair23 commented on hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/HandleContext.java:
```diff
@@ -58,4 +63,29 @@ public interface HandleContext {
+     * @return the verification for the given key, or {@code null} if no such key was provided during pre-handle
+     */
+    @Nullable
+    SignatureVerification verificationFor(@NonNull final Key key);
+
```
Question for @tinker-michaelj : Do we ever need the ability, from the transaction handlers, to get the verification status for a single key? Or do we just want either the total pass/fail information or the Key for a given hollow account?

tinker-michaelj replied:
The schedule service and (for at least some time) the contract service will both need the list of cryptographic keys with valid signatures. 

But I can't immediately think of a case where handler needs to re-check if a Hedera `Key` has an active signature.

rbair23 replied:
@tinker-michaelj and I discussed this question this morning, and we concluded that the method signature is OK (we don't need to be able to get a list of keys), but the specification is wrong. If somebody passes a cryptographic key (i.e. a primitive `Key` such as ed25519 / ecdsa / etc) then we should get back the `SignatureVerification` for that key. But if they pass in a KeyList or ThresholdKey (or a whole hierarchy of them), then we should also return a `SignatureVerification` that represents pass/fail for that whole structure.

hendrikebbers commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/hapi/SignatureVerificationImpl.java:
```diff
@@ -0,0 +1,73 @@
+     * Create a {@link SignatureVerificationImpl} because the given {@link Key} was invalid (did not pass signature
+     * check).
+     */
+    public static SignatureVerification invalid(@NonNull Key key) {
```
NonNull for return values is missing in this class

hendrikebbers commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/AdaptedMonoProcessLogic.java:
```diff
@@ -58,29 +71,80 @@ public void incorporateConsensusTxn(final ConsensusTransaction platformTxn, fina
+        try {
+            return JKey.mapKey(key);
+        } catch (DecoderException e) {
+            throw new RuntimeException("Unable to map key", e);
```
IllegalArgumentException instead?

hendrikebbers commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/AdaptedMonoProcessLogic.java:
```diff
@@ -58,29 +71,80 @@ public void incorporateConsensusTxn(final ConsensusTransaction platformTxn, fina
+            throw new RuntimeException("Unknown failure", e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException("Interrupted while handling sigs", e);
```
IllegalStateException instead?

hendrikebbers commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/AdaptedMonoProcessLogic.java:
```diff
@@ -58,29 +71,80 @@ public void incorporateConsensusTxn(final ConsensusTransaction platformTxn, fina
+                        return future.get(3, TimeUnit.MINUTES);
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        throw new RuntimeException(e);
```
Can you add a message to the exceptions?

hendrikebbers commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/AdaptedMonoProcessLogic.java:
```diff
@@ -58,29 +71,80 @@ public void incorporateConsensusTxn(final ConsensusTransaction platformTxn, fina
+        return map.values().stream()
+                .map(future -> {
+                    try {
+                        return future.get(3, TimeUnit.MINUTES);
```
How is that duration defined? should we make it configurable (by a property)?

hendrikebbers commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:
```diff
@@ -54,192 +57,211 @@
+                task.platformTx.setMetadata(result);
+            } catch (InterruptedException e) {
+                // The thread should only ever be interrupted during shutdown, so we can just log the error.
+                logger.error("Interrupted while waiting for a transaction to be pre handled.");
```
why do you not add the exception in that case?

rbair23 replied:
Oops!

hendrikebbers commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:
```diff
@@ -54,192 +57,211 @@
+            // The node SHOULD have verified the transaction before it was submitted to the network.
+            // Since it didn't, it has failed in its due diligence and will be charged accordingly.
+            return nodeDueDiligenceFailure(creator, preCheck.responseCode(), null);
+        } catch (Throwable th) {
```
Should this case be logged?

rbair23 replied:
I think you are exactly right. This should NEVER happen, and if it does, having it show up in the logs would be super helpful.

jsync-swirlds replied:
Better, don't catch Throwable here.  Any exception thrown will be processed as an ExecutionException or, if the suggestion above using CompletableFutures is used, will be passed to the consumer accept method and be handled (and logged) there.

hendrikebbers commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:
```diff
@@ -54,192 +57,211 @@
-            dispatcher.dispatchPreHandle(context);
+        // 2. Get Payer Account
+        final var txId = txInfo.txBody().transactionID();
+        assert txId != null : "TransactionID should never be null, transactionChecker forbids it";
```
Do we want to have `assert` in runtime code?

hendrikebbers commented on hedera-node/hedera-app/src/testFixtures/java/module-info.java:
```diff
@@ -0,0 +1,11 @@
+    requires com.hedera.node.hapi;
+    requires com.hedera.pbj.runtime;
+    requires org.assertj.core;
+    requires com.github.spotbugs.annotations;
```
```suggestion
    requires static com.github.spotbugs.annotations;
```

netopyr commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/hapi/SignatureVerificationFuture.java:
```diff
@@ -0,0 +1,305 @@
+            while (!futureReady) {
+                // Figure out how many milliseconds we still have remaining before timing out
+                final var now = System.currentTimeMillis();
+                millisRemaining -= System.currentTimeMillis() - now;
```
This does not look right.

netopyr commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/ReceiptCache.java:
```diff
@@ -0,0 +1,89 @@
+     * @param nodeAccountID The node that has seen this transaction
+     */
+    /*@ThreadSafe*/
+    void record(@NonNull TransactionID transactionID, @NonNull AccountID nodeAccountID);
```
Maybe we should avoid `record` as a method name from now on.

netopyr commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/ReceiptCache.java:
```diff
@@ -0,0 +1,89 @@
+    /**
+     * Represents a single item in the cache.
+     * @param transactionID The transaction ID associated with the receipt
+     * @param nodeAccountIDs The list nodes which have independently submitted events with the same transaction ID
```
I am a little confused. Here we are talking about nodes, that have **submitted** events with the same transaction ID. Below we record nodes that have **seen** a transaction ID. Is that the same thing?

rbair23 replied:
This will be redone in #6555

tinker-michaelj commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/hapi/SignatureVerificationFuture.java:
```diff
@@ -0,0 +1,305 @@
+ * A {@link Future} that waits on a {@link List} of {@link TransactionSignature}s to complete signature checks, and
+ * yields a {@link SignatureVerification}.
+ */
+class SignatureVerificationFuture implements Future<SignatureVerification> {
```
Have we considered using `CompletableFuture` machinery here? It feels like we could get the necessary behavior out-of-the-box.

rbair23 replied:
I tried, but didn't find a way in any JDK classes to aggregate multiple futures into one. And then after talking with you I realized that determining "pass/fail" requires walking over the key structure again and comparing which signatures passed/failed to determine whether, for example, a ThresholdKey is passing or not. I'm not sure what to do about it.

tinker-michaelj commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/hapi/SignatureVerificationFuture.java:
```diff
@@ -0,0 +1,305 @@
+     * @return {@code true} if at least {@link ThresholdKey#threshold()} keys within the {@code key} have been
+     * validated, {@code false} otherwise.
+     */
+    private boolean checkIfPassed(@NonNull final KeyList keyList, @NonNull final Map<Key, TransactionSignature> map) {
```
There is a special case that if we are checking the signature for a file's `KeyList` during `FileDelete`, we need to pass this check if _any_ key in the list signs (c.f. [here](https://github.com/hashgraph/hedera-protobufs/blob/main/services/file_delete.proto#L32)).

rbair23 replied:
I saw that comment, but it seems very surprising (and wrong!). Why do we do that? One disgruntled signer can nuke a file?

rbair23 replied:
Maybe what the `FileDelete` will need to do is shred the keylists. Basically, given a `k1`, create a new key `k2`, such that all key lists are converted into threshold keys with a threshold of 1.

rbair23 commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/hapi/SignatureVerificationFuture.java:
```diff
@@ -0,0 +1,305 @@
+    /**
+     * {@inheritDoc}
+     *
+     * Attempts to cancel all {@link TransactionSignature} futures. Due to the asynchronous nature of those objects,
```
I agree, I found `TransactionSignature` to be a bit strange in this way.

tinker-michaelj commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/hapi/SignatureVerifierImpl.java:
```diff
@@ -0,0 +1,387 @@
+                // to uncompress it first.
+                final var compressedPrefixByteArray = new byte[(int) prefix.length()];
+                prefix.getBytes(0, compressedPrefixByteArray);
+                // It may be that the bytes are random nonsense, in which case decompressing will throw an exception.
```
Maybe,
```
If the compressed key begins with a prefix other than `0x00`-`0x07`, decompressing will throw an exception
```

tinker-michaelj commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/hapi/SignatureVerificationFuture.java:
```diff
@@ -0,0 +1,305 @@
+     * @return {@code true} if the {@code key} has been validated, {@code false} otherwise.
+     */
+    private boolean checkIfPassed(@NonNull final Key key, @NonNull final Map<Key, TransactionSignature> map) {
+        return switch (key.key().kind()) {
```
These three mutually recursive `check*()` variants will be convenient to extract to static methods that accept a strategy interface for testing if any of the "base case" kinds of `Keys`s have valid signatures.

rbair23 replied:
If you are OK I'd like to leave this as an exercise for a future PR. I tried to factor this out several times (similar logic is also in the SignatureVerifierImpl), but I couldn't see a nice way to do it. The logic between those two places was slightly different. I thought about using a Visitor pattern approach, but somebody on the internet wisely remarked that the visitor pattern and switch expressions do the same thing but switch expressions are way less overhead!

tinker-michaelj commented on hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/HandleContext.java:
```diff
@@ -58,4 +63,29 @@ public interface HandleContext {
+     * @param hollowAccountID the hollow account ID to get the verification for
+     * @return the verification for the given account, or {@code null} if no such account was provided during pre-handle
+     */
+    @Nullable
+    SignatureVerification verificationFor(@NonNull final AccountID hollowAccountID);
```
In `mono-service` the workflow decides whether to do hollow account completions before calling a `handler()` at all. So if only handlers were going to use this API, it's probably not needed.

Not certain though.

rbair23 replied:
OK, I will leave it in for now, but we may not need it.

tinker-michaelj commented on hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java:
```diff
@@ -102,10 +107,8 @@ private PreHandleContext(
         // Find the account, which must exist or throw a PreCheckException with the given response code.
         final var account = accountAccess.getAccountById(payer);
         mustExist(account, responseCode);
-        // NOTE: While it is true that the key can be null on some special accounts like
-        // account 800, those accounts cannot be the payer.
+        // It is possible for the payer key to be null if the payer is a hollow account!
         this.payerKey = account.key();
-        mustExist(this.payerKey, responseCode);
     }
```
In `mono-service` hollow accounts also have an empty `KeyList`. Might just be an implementation detail?

rbair23 replied:
Is it something we should mandate? In this case, the test code was not configured with a payer key, but with null as the key. We could put this check back in, but then update the tests to use an empty key list. Would that be better?

rbair23 replied:
I've reverted this so payer cannot be null with the assumption that hollow accounts will have an empty keylist

rbair23 commented on hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/ManualExecutorService.java:
```diff
@@ -0,0 +1,78 @@
+    private boolean shutdown = false;
+
+    public void runAllTasks() {
+        while (!tasks.isEmpty()) {
```
TBH this class isn't even used right now. Why don't I just remove it and we can add it back in later if it is useful

rbair23 commented on hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/ImmediateExecutorService.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Override
+    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+        return true;
```
You are right. I didn't bother implementing this method since I wasn't using it, but I have implemented it now, I think correctly.

rbair23 commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/hapi/SignatureVerifierImpl.java:
```diff
@@ -0,0 +1,387 @@
+
+        // There may be duplicate keys in the above list. We need to remove them. (It would be a waste of resources
+        // to do duplicate signature verifications).
+        final var txSigs = list.stream().distinct().toList();
```
I would not be surprised if it turned out to be horrible. I had a hand-written two dimensional loop instead, which I expect to be faster but was harder to reason about. If it shows up in the profiles, we'll replace this. I really tried to find some better way to remove duplicates, but they were all bad. In the end, I think the two dimensional for loop is the best option of all the bad, if this turns out to be a problem.

rbair23 commented on hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/workflows/PreHandleContext.java:
```diff
@@ -193,6 +201,31 @@ public PreHandleContext requireKey(@NonNull final Key key) {
+                || hollowAccount.alias() == null
+                || hollowAccount.alias().length() != 20) {
+            throw new IllegalArgumentException("Account " + hollowAccount.accountNumber() + " is not a hollow account");
+        }
```
I've consolidated these checks into HapiUtils.

netopyr commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:
```diff
@@ -54,192 +57,208 @@
-        CompletableFuture.allOf(array).join();
+        // Waits for all the background threads to complete their work and stores the resulting PreHandleResult
+        // as the transaction's metadata.
+        for (final var task : tasks) {
```
I would have preferred to do everything in a single loop, but it is probably just a matter of personal taste.

jsync-swirlds replied:
It could be done in a single loop using CompletableFuture and using the fluent API for that class to apply the result process on each, then just join them all at the end.  As a bonus you get a single timeout, which further simplifies things and reduces possible errors.
e.g. (note, indentation may be off; typing code in github comments is not ideal; spotless probably won't like things here; there may be bugs)
```
        var tasks = new ArrayList<CompletableFuture>(512); // generally not much benefit going over 512 due to the array of arrays ArrayList uses.
        while (transactions.hasNext()) {
            // Skip platform (system) transactions.
            final var platformTx = transactions.next();
            if (platformTx.isSystem()) continue;

            // Submit the task with its completion process to put the result as the metadata on the transaction to the executor service
            final var future =
                    CompletableFuture.runAsync(() ->
                    preHandleTransaction(creator, readableStoreFactory, accountStore, platformTx), exe)
                    .whenCompleteAsync(new PreHandleResultHandler(platformTx), exe);
            tasks.add(future);
        }
        final var config = configProvider.getConfiguration().getConfigData(PreHandleConfig.class);
        final var timeout = config.timeout();
        CompletableFuture.allOf(tasks).orTimeout(timeout).join();
        // at this point all futures are complete and each has completed the code below
        ...
   }
...
private static class PreHandleResultHandler implements BiConsumer<PreHandleResult, Throwable> {
    private final platformTransaction;

    public PreHandleResultHandler(@NonNull final Transaction platformTx) {
        platformTransaction = Objects.requiresNonNull(platformTx);
    }

    @Override
    public void accept(final PreHandleResult result, final Throwable verificationFailure) {
        if(verificationFailure == null) {
            platformTransaction.setMetadata(result);
        } else {
            switch(verificationFailure) {
                case InterruptedException e -> {
                    // The thread should only ever be interrupted during shutdown, so we can just log the error.
                    logger.error("Interrupted while waiting for a transaction to be pre handled.", e);
                    Thread.currentThread().interrupt();
                    task.platformTx.setMetadata(unknownFailure());
Member
                }
                case ExecutionException e -> {
                    logger.error("Unexpected error while pre handling a transaction!", e);
                    task.platformTx.setMetadata(unknownFailure());
                }
                case TimeoutException e -> {
                    logger.error("Timed out while waiting for a transaction to be pre handled!", e);
                    task.platformTx.setMetadata(unknownFailure());
Member
                }
            }
        }
    }
}
```

rbair23 replied:
I wanted to acknowledge before closing this conversation that I have seen this and appreciate it. I didn't see it before because this section of code was redone based on a review by @jasperpotts before I got to the comment but I really appreciate the time to write this up and I learned a few things along the way!

netopyr commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:
```diff
@@ -54,192 +57,208 @@
+        // 4. Create the PreHandleContext. This will get reused across several calls to the transaction handlers
+        final PreHandleContext context;
+        try {
+            // NOTE: Once PreHandleContext is moved from being a concrete implementation in SPI, to being an Interface/
```
Well, `PreHandleContext` has already become an interface... 🙂

rbair23 replied:
Honestly I would fix this comment right now but if I do I lose all my approvals on the PR 😡 .

Neeharika-Sompalli commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleResult.java:
```diff
@@ -16,76 +16,122 @@
+    /**
+     * An enumeration of all possible types of pre-handle results.
+     */
+    public enum Status {
```
Are we adding this Status in addition to `ResponseCodeEnum` so we can easily check in `handle` and throw away if its `PRE_HANDLE_FAILURE` ?

Neeharika-Sompalli commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:
```diff
@@ -54,192 +57,208 @@
+                // The thread should only ever be interrupted during shutdown, so we can just log the error.
+                logger.error("Interrupted while waiting for a transaction to be pre handled.", e);
+                Thread.currentThread().interrupt();
+                task.platformTx.setMetadata(unknownFailure());
```
Should we have a different failure for `InterruptedException` because its not UNKNOWN ? Does this also mean we return UNKNOWN error code in `HandleException` also whenever we receive `InterruptedException`? 
I personally don't like `UNKNOWN` since it doesn't provide any info

rbair23 replied:
I think we just need a status code that indicates that "something bad happened, it doesn't really matter what, try again later". I'm using UNKNOWN for that, but maybe there is a better name for it. But I don't think it is useful to disambiguate what exactly happened (InterruptedException vs. some other Throwable) unless we're going to take distinct action for those different types. At the moment I don't think we will (?) but if we do, I agree it should be clarified here.

Neeharika-Sompalli commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:
```diff
@@ -54,192 +57,208 @@
+                Thread.currentThread().interrupt();
+                task.platformTx.setMetadata(unknownFailure());
+            } catch (ExecutionException e) {
+                logger.error("Unexpected error while pre handling a transaction!", e);
+                task.platformTx.setMetadata(unknownFailure());
+            } catch (TimeoutException e) {
+                logger.error("Timed out while waiting for a transaction to be pre handled!", e);
+                task.platformTx.setMetadata(unknownFailure());
```
Same for these . Should we return `UNKNOWN` response ?

rbair23 reviewed (commented):
LGTM

jsync-swirlds commented on hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/HandleContext.java:
```diff
@@ -71,4 +74,28 @@ public interface HandleContext {
+     * @return the verification for the given key, or {@code null} if no known key or signature at pre-handle
+     */
+    @Nullable
+    SignatureVerification verificationFor(@NonNull final Key key);
```
Note (probably for a future PR):
If we keep this single verification approach, the method the signature verification code uses to walk the tree should be exposed in a utility, so other code (e.g. ScheduleSign) that needs to take a set of keys and decompose that into a flat list of many possibly optional cryptographic keys can do so without duplicating the same complex code.
For example:
Schedule Sign dispatches a pre-handle for a Token Transfer internal transaction that requires 8 signatures, two are keylist, three are MofN threhsolds, and three are single keys.  ScheduleSign then needs to get all of the individual keys (all the way down to the cryptographic keys under those "keys") and add them as optional keys on the context because any subset of individual cryptographic keys might have signed the current transaction.

Rather than recreate the code to walk a potentially extensive grove of trees of keys and subkeys in Schedule Service, it would be good to have a utility method that can be called to get all of the individual keys that is then shared from the verification code.

rbair23 replied:
I agree, we need to find a way to share that logic. Good thing for follow up PR.

jsync-swirlds commented on hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/ImmediateExecutorService.java:
```diff
@@ -0,0 +1,86 @@
+
+    @Override
+    public synchronized List<Runnable> shutdownNow() {
+        // Note, since this method is synchronized, it is not possible for another thread to concurrently add a new
```
There is no queue to be modified concurrently, so likely this method (and all the others) doesn't need to be synchronized at all (particularly if the shutdown boolean on line 33 is declared volatile).

jsync-swirlds commented on hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/ImmediateExecutorService.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @Override
+    public synchronized void execute(@NonNull final Runnable command) {
```
Given that we run immediately instead of submitting for use on another thread, do we really need this to be synchronized?

jsync-swirlds commented on hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/ImmediateExecutorService.java:
```diff
@@ -0,0 +1,86 @@
+
+/**
+ * An {@link ExecutorService} that executes tasks immediately in the calling thread, rather than in a separate thread.
+ * This is useful for synchronous testing where the API involves an {@link ExecutorService}.
```
The requirement of everything to be synchronized may slow down testing quite substantially (not sure).  Based on how the AbstractExecutorService (which has no synchronized methods) calls execute inline rather than managing any queue itself, I believe we could implement this correctly without synchronized by declaring shutdown to be volatile.
The only point of potential issue I see would be if we have calls to InvokeAll, which _might_ still complete one or more tasks from that set on its calling thread after another thread called shutdown.
In a testing context, I don't believe that would be significant (though we should verify that).

rbair23 replied:
So far I haven't seen any slowdown in the tests. You may be right that we don't need to use synchronized, but doing so makes it obviously correct, whereas without it, I would need to devote more brainpower and validation to it to make sure it is correct. At this point, I don't think the payoff is worth it. But we could improve this in a follow on PR if it is a problem (but I really don't think it will be, since synchronization is really very fast in wall clock terms, it is only a problem under contention, and in this case we should almost never have contention since most tests are single threaded).

jsync-swirlds commented on hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/ImmediateExecutorService.java:
```diff
@@ -0,0 +1,86 @@
+    public synchronized List<Runnable> shutdownNow() {
+        // Note, since this method is synchronized, it is not possible for another thread to concurrently add a new
+        // task at the same time. And this method cannot be called while another thread is executing a task. So we can
+        // safely return an empty list.
```
I believe we can return an empty list regardless, as there is no queue of tasks and no set of active threads that this code could track.
For reference, when the rejected execution policy is caller runs, the shutdownNow method of the standard classes also doesn't return those tasks that are running on the caller thread; only tasks in the active queue are returned.

rbair23 replied:
I think you're right. This implementation would just assume that there is nothing in the queue (since there is no queue).

rbair23 replied:
Do you want me to remove the comment? The code itself wouldn't change. Not sure if it is worth resetting approvals :)

jsync-swirlds replied:
There will be changes somewhere; just add this to the list to do then.

hendrikebbers commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/TransactionChecker.java:
```diff
@@ -452,4 +472,62 @@ private <T extends Record> T parseStrict(
+     * @param sigPairs The list of signature pairs to sort. Cannot be null.
+     * @return the sorted list of signature pairs
+     */
+    private List<SignaturePair> sort(@NonNull final List<SignaturePair> sigPairs) {
```
```suggestion
    @NonNull private List<SignaturePair> sort(@NonNull final List<SignaturePair> sigPairs) {
```

hendrikebbers commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/CompoundSignatureVerificationFuture.java:
```diff
@@ -0,0 +1,218 @@
+            final int numCanFail) {
+        this.key = requireNonNull(key);
+        this.hollowAccount = hollowAccount;
+        this.futures = requireNonNull(futures);
```
Maybe we should copy the list to an immutable list. Otherwise the `if (futures.isEmpty()) {` check at the end of the constructor is worthless since the list could be mutated from the outside.

rbair23 replied:
We could, and I definitely appreciate the safety argument. My only concern is garbage generation. The fact that `CompoundSignatureVerificationFuture` is package private and only used by `PreHandleResult` in one place where `PreHandleResult` creates the array list as a local variable in the function makes me think this level of (very very good) paranoia is maybe not needed in this case. I myself am very paranoid about code, and I like the idea of being very defensive. But I worry about all the object allocation too. I wish there was a better way to prove immutability of the list and throw an exception if it was ever a mutable list.

hendrikebbers commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/CompoundSignatureVerificationFuture.java:
```diff
@@ -0,0 +1,218 @@
+                return new SignatureVerificationImpl(key, hollowAccount, false);
+            }
+            // Figure out how many milliseconds we still have remaining before timing out
+            millisRemaining -= System.currentTimeMillis() - beforeWaiting;
```
I would add a check for `millisRemaining <= 0` and throw a `TimeoutException` in that case.

rbair23 replied:
Using @jsync-swirlds's approach to the other variant of this, what if I had this implementation:

```
    @Override
    public SignatureVerification get(final long timeout, final TimeUnit unit)
            throws InterruptedException, ExecutionException, TimeoutException {
        int failCount = 0;
        final var deadline = System.currentTimeMillis() + unit.toMillis(timeout);
        for (final var future : futures) {
            // We know the maximum number of millis we can wait, so we block for at most that long. If it takes
            // longer, the future we delegate to here will throw. If it takes less, we recompute how much is
            // remaining on the next iteration.
            final var now = System.currentTimeMillis();
            final var verification = future.get(deadline - now, TimeUnit.MILLISECONDS);
            if (!verification.passed() && (failCount++ >= numCanFail)) {
                return new SignatureVerificationImpl(key, hollowAccount, false);
            }
        }

        return new SignatureVerificationImpl(key, hollowAccount, true);
    }
```

This gets rid of that line entirely.

hendrikebbers commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowModule.java:
```diff
@@ -45,6 +52,23 @@ static Function<SignatureMap, PubKeyToSigBytes> provideKeyToSigFactory() {
+    SignatureExpander bindSignatureExpander(SignatureExpanderImpl signatureExpander);
+
+    // FUTURE: This should be updated to use the new config module and real config loading when possible
+    @Provides
```
https://github.com/hashgraph/hedera-services/pull/6291 does it. Maybe you can deprecate your call

jsync-swirlds commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/SignatureVerificationFuture.java:
```diff
@@ -0,0 +1,46 @@
+ * A {@link Future} that waits on a {@link Map} of {@link TransactionSignature}s to complete signature checks, and
+ * yields a {@link SignatureVerification}.
+ */
+public interface SignatureVerificationFuture extends Future<SignatureVerification> {
```
Given that there is a clear completion condition, would this be better extending `CompletableFuture`?

rbair23 replied:
It doesn't work because I don't have control over when it completes. It is up to the `TransactionSignature` object's `Future` as to when completion happens.

jsync-swirlds commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:
```diff
@@ -54,192 +61,217 @@
+                // unit tests, so they use a much shorter configuration, so test execution is faster).
+                final var config = configProvider.getConfiguration().getConfigData(PreHandleConfig.class);
+                final var timeout = config.timeout();
+                final var result = task.future.get(timeout.toMillis(), TimeUnit.MILLISECONDS);
```
This could potentially result in waiting many times the timeout if several items are delayed close to, but not over, the timeout limit.
Rather than calling get with a timeout in a loop, we might be better off joining all of the threads with a timeout, then get all the results knowing they're all immediately available.
Changes needed:
 * change the Futures to be CompletableFutures
 * Use `CompletableFuture.allOf(arrayOfFutures).get()` (returns void) to wait for all to complete before entering the loop on line 150; when that call returns all of the futures are guaranteed to have results available (or throw ExecutionException).
 * Adjust to ensure the methods that return Futures do not return null (null causes issues with allOf).
Note, you could also use `CompletableFuture.allOf(arrayOfFutures).thenApplyAsync(this::methodHandleToHandleEachFuture)` to run the loop for items as they arrive (rather than once all are ready).

rbair23 replied:
The problem is the future is not mine, it is created by the `Executor`, so I cannot complete it. I wish we had `allOf` for regular futures :-(.

jsync-swirlds replied:
You can pass the executor to the CompletableFuture call, which avoids the Executor creating a regular Future.  See the example code I wrote above.

rbair23 replied:
Oh, this is very interesting. I did not know this. I'm not sure how I feel about it yet. I have two reactions:

1. We should use `CompletableFuture` everywhere we use Future because `CompletableFuture` has all this awesome API (`join`, `runAfterBoth`, ... and a million other things) by virtue of `CompletionStage`.
2. `CompletableFuture` is the wrong abstraction. The Java team should have made `Future` extend `CompletionStage`

The problem with `CompletableFuture` is it *does* expose API for completing the future. It seems like a bad idea for object 1 to create a future, give it to object 2, and allow object 2 to complete the future, irrespective of whatever background work was actually going on, as the following test demonstrates:

```
    @Test
    void completableFutureOverride() throws ExecutionException, InterruptedException {
        final var exe = Executors.newSingleThreadExecutor();
        final var cf = CompletableFuture.supplyAsync(() -> {
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
            return 1; // The actual value computed by the background thread
        }, exe);

        System.out.println(cf.complete(2)); // prints "true", and supplies a different value than the computing thread
        System.out.println(cf.get()); // prints '2'. It overrode whatever the background thread was doing.
    }
```

That seems wrong and could be dangerous in a program. Sometimes this might be useful and desirable behavior, when you really *want* the ability to override whatever computation was happening in the background. It seems like the only reason for using `CompletableFuture` is because `Future` doesn't extend from `CompletionStage`. Maybe we should have our own `Future` subclass that does, and use that everywhere (it could even be implemented with `CompletableFuture` behind the scenes but hide the dangerous parts).

jsync-swirlds commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/CompoundSignatureVerificationFuture.java:
```diff
@@ -0,0 +1,218 @@
+            throws InterruptedException, ExecutionException, TimeoutException {
+        var millisRemaining = unit.toMillis(timeout);
+        int failCount = 0;
+        for (final var future : futures) {
```
Might be good to use CompletableFuture.allOf to set a single timeout, then process all of the items (counting success and failure for your failCount entries)
```
// wait for all to complete, or timeout
CompletableFuture.allOf(arrayOfCompletableFutures).orTimeout(timeout, unit).join();
// process each future in the array
```

jsync-swirlds replied:
A CompletableFuture is complete when it's value is available, it's not generally completed by code outside the task it represents.  I would claim that any Future that returns a single result value can be considered a CompletableFuture.

jsync-swirlds commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/impl/SignatureVerificationFutureImpl.java:
```diff
@@ -0,0 +1,192 @@
+            final var now = System.currentTimeMillis();
+            millisRemaining -= (now - time);
+            if (millisRemaining <= 0) {
+                // If there was no time left, then TimeoutException
```
Possibly cleaner rewrite:
```
        final long deadline = System.currentTimeMillis() + unit.toMillis(timeout);
        while (txSig.getFuture() == null) {
            if (System.currentTimeMillis() > deadline) {

```
Further down
```
        txSig.getFuture().get(deadline - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
```

rbair23 replied:
I love it. Will integrate and push.

rbair23 commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java:
```diff
@@ -54,192 +61,217 @@
     @Inject
     public PreHandleWorkflowImpl(
+            @NonNull final ConfigProvider configProvider,
             @NonNull final ExecutorService exe,
```
I think I actually need the `ExecutorService` because `Executor` only runs `Runnables` but I need a `Callable` since it has a return type.

rbair23 commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/impl/SignatureVerificationFutureImpl.java:
```diff
@@ -0,0 +1,192 @@
+        canceled = true;
+        final var future = txSig.getFuture();
+        if (future != null) {
+            return future.cancel(mayInterruptIfRunning);
```
I am not sure that is right. The API on Future#cancel says:

> Attempts to cancel execution of this task. This method has no effect if the task is already completed or cancelled, or could not be cancelled for some other reason. Otherwise, if this task has not started when cancel is called, this task should never run. If the task has already started, then the mayInterruptIfRunning parameter determines whether the thread executing this task (when known by the implementation) is interrupted in an attempt to stop the task.

>The return value from this method does not necessarily indicate whether the task is now cancelled; use isCancelled.

And the definition of the return value is:

>false if the task could not be cancelled, typically because it has already completed; true otherwise. If two or more threads cause a task to be cancelled, then at least one of them returns true. Implementations may provide stronger guarantees.

If the underlying future doesn't exist yet, then as far as I'm concerned it will never exist (i.e. if it ever does exist, the caller of `SignatureVerificationFutureImpl` doesn't ever need to know that, as far as they are concerned, this future is canceled). If the underlying future does exist, then I will return whatever it reports, but that doesn't impact whether or not this future is canceled. It is canceled either way. That is how I understand the specification on Future.

rbair23 commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/impl/SignatureVerificationFutureImpl.java:
```diff
@@ -0,0 +1,192 @@
+                Thread.sleep(1);
+            }
+        }
+        txSig.getFuture().get(millisRemaining, TimeUnit.MILLISECONDS); // Wait for the future to complete
```
You are right. I rewrote it per @jsync-swirlds suggestion:

```
    public SignatureVerification get(long timeout, TimeUnit unit)
            throws InterruptedException, ExecutionException, TimeoutException {
        final long deadline = System.currentTimeMillis() + unit.toMillis(timeout);
        while (txSig.getFuture() == null) {
            if (System.currentTimeMillis() > deadline) {
                // If there was no time left, then TimeoutException
                throw new TimeoutException("Timed out waiting for signature verification to complete");
            } else {
                // If there was time left, but we didn't yet have a Future on the TransactionSignature, then
                // rather than blocking in a tight loop (like we do for "get"), we have to just wait a bit and
                // check for timeout and try again. That way, we can still time out.
                //noinspection BusyWait
                Thread.sleep(1);
            }
        }

        // Wait for the future to complete
        txSig.getFuture().get(deadline - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
        return new SignatureVerificationImpl(key, hollowAccount, txSig.getSignatureStatus() == VALID);
    }
```

This seems simpler and I think addresses the concern.

rbair23 commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/CompoundSignatureVerificationFuture.java:
```diff
@@ -0,0 +1,218 @@
+        // Try to cancel each underlying future (I go ahead and try canceling all of them, even if one fails).
+        // If the underlying Future doesn't exist, then we just skip it. Best effort.
+        for (final var future : futures) {
+            future.cancel(mayInterruptIfRunning);
```
Seems like if any one of them returns false we should return false, otherwise return true.

rbair23 commented:
I have addressed all open comments, fixed all the ones I felt were right. The other comments are still open so we can continue dialog if needed. Please let met know if we're OK to proceed.

I fixed the bugs we identified in the recorded call today. I also made a change to how the threading was handled in `PreHandleWorkflowImpl` based on a live review with Jasper. I decided to pull out the executor service and just use the fork join thread pool. The code is way, way simpler.

tinker-michaelj reviewed (approved):
LGTM!

Neeharika-Sompalli reviewed (approved):
LGTM !

Neeharika-Sompalli commented on hedera-node/hedera-app-spi/src/testFixtures/java/com/hedera/node/app/spi/fixtures/TestBase.java:
```diff
@@ -83,6 +83,21 @@ public final Bytes randomBytes(int length) {
+     * @return Some random bytes.
+     */
+    @NonNull
+    public final byte[] randomByteArray(int length) {
```
nit : should we probably use `randomBytes(rand, length)` inside this ?

rbair23 replied:
I would apply this change but if I do I lose my approvers 😡 . Next time!

rbair23 commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/hapi/SignatureVerificationFuture.java:
```diff
@@ -0,0 +1,304 @@
+                    // If there was time left, but we didn't yet have a Future on the TransactionSignature, then
+                    // rather than blocking in a tight loop (like we do for "get"), we have to just wait a bit and
+                    // check for timeout and try again. That way, we can still time out.
+                    Thread.sleep(1);
```
Yes, `TransactionSignature` needs some love :-)

rbair23 commented on hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/hapi/SignatureVerificationFuture.java:
```diff
@@ -0,0 +1,304 @@
+ * A {@link Future} that waits on a {@link Map} of {@link TransactionSignature}s to complete signature checks, and
+ * yields a {@link SignatureVerification}.
+ */
+class SignatureVerificationFuture implements Future<SignatureVerification> {
```
What I was hoping was that in the public API we just say this is a Future to keep things as decoupled at the API level as we can. So this class just becomes an implementation detail.
